Error thrown on deserialising the feed from cache in getFeed in FeedRepository

2024-04-17T16:12:46.664+05:30 ERROR 11462 --- [onPool-worker-1] org.araa.feed.FeedService                : Error fetching feed for https://www.newswire.lk/feed

java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.araa.feed.Feed cannot be cast to class org.araa.feed.Feed (org.araa.feed.Feed is in unnamed module of loader 'app'; org.araa.feed.Feed is in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @532504fb)
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165) ~[na:na]
Caused by: java.lang.ClassCastException: class org.araa.feed.Feed cannot be cast to class org.araa.feed.Feed (org.araa.feed.Feed is in unnamed module of loader 'app'; org.araa.feed.Feed is in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @532504fb)
	at org.araa.feed.FeedRepositoryImp$$SpringCGLIB$$0.getFeed(<generated>) ~[na:na]
	at org.araa.feed.FeedService.lambda$fetchFeed$0(FeedService.java:18) ~[classes/:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) ~[na:na]
	... 6 common frames omitted

--- a/src/main/java/org/araa/application/dto/Feed.java
+++ b/src/main/java/org/araa/application/dto/Feed.java
@@ -7,12 +7,13 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.araa.domain.Entry;
 
+import java.io.Serializable;
 import java.util.List;
 
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
-public class Feed {
+public class Feed implements Serializable {
     /**
      * This is wrapper class for SyndFeed
      * It is used to extract the required fields from SyndFeed

--- a/src/main/java/org/araa/domain/Category.java
+++ b/src/main/java/org/araa/domain/Category.java
@@ -5,13 +5,14 @@ import lombok.AllArgsConstructor;
 import lombok.Data;
 import lombok.NoArgsConstructor;
 
+import java.io.Serializable;
 import java.util.Date;
 
 @NoArgsConstructor
 @AllArgsConstructor
 @Data
 @Entity
-public class Category {
+public class Category implements Serializable {
     @Id
     @GeneratedValue( strategy = GenerationType.IDENTITY )
     @Column( name = "category_id" )

--- a/src/main/java/org/araa/domain/Entry.java
+++ b/src/main/java/org/araa/domain/Entry.java
@@ -1,18 +1,16 @@
 package org.araa.domain;
 
 import jakarta.persistence.*;
-import lombok.AllArgsConstructor;
 import lombok.Data;
-import lombok.NoArgsConstructor;
 
+import java.io.Serializable;
 import java.util.Date;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Set;
 
 @Data
 @Entity
-public class Entry {
+public class Entry implements Serializable {
     /**
      * This is wrapper class for SyndEntry
      * It is used to extract the required fields from SyndEntry

--- a/src/main/java/org/araa/domain/RSS.java
+++ b/src/main/java/org/araa/domain/RSS.java
@@ -3,12 +3,13 @@ package org.araa.domain;
 import jakarta.persistence.*;
 import lombok.Data;
 
+import java.io.Serializable;
 import java.util.Date;
 
 @Entity
 @Data
 @Table( name = "rss" )
-public class RSS {
+public class RSS implements Serializable {
 
     @Id
     @GeneratedValue( strategy = GenerationType.IDENTITY )

--- a/src/main/java/org/araa/infrastructure/config/RedisCacheConfig.java
+++ b/src/main/java/org/araa/infrastructure/config/RedisCacheConfig.java
@@ -6,7 +6,7 @@ import org.springframework.context.annotation.Configuration;
 import org.springframework.data.redis.cache.RedisCacheConfiguration;
 import org.springframework.data.redis.cache.RedisCacheManager;
 import org.springframework.data.redis.connection.RedisConnectionFactory;
-import org.springframework.data.redis.serializer.GenericJackson2JsonRedisSerializer;
+import org.springframework.data.redis.serializer.JdkSerializationRedisSerializer;
 import org.springframework.data.redis.serializer.RedisSerializationContext;
 
 import java.time.Duration;
@@ -17,9 +17,9 @@ public class RedisCacheConfig {
 
     @Bean( name = "redisCacheManager" )
     public RedisCacheManager cacheManager( RedisConnectionFactory connectionFactory ) {
-        RedisCacheConfiguration cacheConfiguration = RedisCacheConfiguration.defaultCacheConfig( Thread.currentThread().getContextClassLoader() )
+        RedisCacheConfiguration cacheConfiguration = RedisCacheConfiguration.defaultCacheConfig()
                 .entryTtl( Duration.ofMinutes( 60 ) )
-                .serializeValuesWith( RedisSerializationContext.SerializationPair.fromSerializer( new GenericJackson2JsonRedisSerializer() ) );
+                .serializeValuesWith( RedisSerializationContext.SerializationPair.fromSerializer( new JdkSerializationRedisSerializer( Thread.currentThread().getContextClassLoader() ) ) );
 
         return RedisCacheManager.builder( connectionFactory )
                 .cacheDefaults( cacheConfiguration )

--- a/src/main/java/org/araa/repositories/FeedRepositoryImp.java
+++ b/src/main/java/org/araa/repositories/FeedRepositoryImp.java
@@ -29,9 +29,6 @@ public class FeedRepositoryImp implements FeedRepository {
 
     @Cacheable( value = "feed", key = "#rss", unless = "#result == null" )
     public Feed getFeed( String rss ) {
-        // this is required to set the correct class loader for the current thread
-        Thread.currentThread().setContextClassLoader( getClass().getClassLoader() );
-
         RedisCache cache = ( RedisCache ) Objects.requireNonNull( cacheManager.getCache( "feed" ) );
         logger.info( "Class loader for Feed: {}", Feed.class.getClassLoader() );
         Feed feed = cache.get( rss, Feed.class );  // This will automatically deserialize the JSON back into a Feed object

